design/FY2018/NFP-Table-Refactor.md

```diff
@@ -0,0 +1,819 @@
+# Refactoring of Performance Tables
+
+## Justification for New Feature
+
+The input of performance tables into EnergyPlus is a crucial part of most building energy modeling efforts, particularly in the representation of the performance of HVAC equipment. Recent feedback indicates that the current facility is increasingly becoming a pain point for users. Among the challenges users face are:
+
+- The EnergyPlus documentation does not always match the implemented capabilities;
+- Incomplete inputs are not fully vetted before computations are made, and
+- Some tables are untested.
+
+ASHRAE Standard 205, which seeks to facilitate sharing of equipment characteristics for performance simulation, is nearing completion, so better support for the upcoming standard is a highly desirable feature and provides an opportunity to address some of the current table-related issues in EnergyPlus. In support of ASHRAE Standard 205 integration into EnergyPlus, a new N-dimensional linear interpolation routine for gridded data is required. This is a critical capability for any tool utilizing ASHRAE 205 equipment data and the improved functionality will remove a noted deficiency.
+
+## Overview
+
+This work will be accomplished through several tasks:
+
+1. Develop an N-dimensional interpolation method for gridded data. Where appropriate, utilization of existing open source routines and/or libraries is permitted provided that prior approval has been explicitly obtained from NREL prior to implementation. The interpolation method is not required to fill in any missing input data, but must recognize missing input data.
+2. Refactor existing EnergyPlus table procedures to utilize the new methods or initiate the deprecation process for capabilities that are not currently fully functional.
+3. Clean up EnergyPlus table input objects to have full, as-advertised, functionality. Where possible, warning and/or error messages should be added to advise the user of incomplete inputs.
+
+This New Feature Proposal focuses largely on tasks #2 and #3. The interpolation method will be further defined in the design document.
+
+The general proposal is to collapse the existing three table objects:
+
+- *Table:OneIndependentVariable*,
+- *Table:TwoIndependentVariables*, and
+- *Table:MultiVariableLookup*
+
+into a single new object: *Table:Lookup*.
+
+## Discussion
+
+### Abandoning curve-fitting capability
```

Comment:
Are we sure we want to drop the curve fitting? Even for 1 and 2 dimensions?

Reply:
I think for now we won't add curve-fitting to the new table. We won't immediately remove the existing tables and their ability to perform regressions.